src/Microsoft.AspNetCore.TestHost/WebSocketClient.cs

```diff
@@ -181,12 +190,20 @@ bool IHttpWebSocketFeature.IsWebSocketRequest
 
             Task<WebSocket> IHttpWebSocketFeature.AcceptAsync(WebSocketAcceptContext context)
             {
-                Context.HttpContext.Response.StatusCode = 101; // Switching Protocols
-
                 var websockets = TestWebSocket.CreatePair(context.SubProtocol);
-                _clientWebSocketTcs.SetResult(websockets.Item1);
-                _serverWebSocket = websockets.Item2;
-                return Task.FromResult<WebSocket>(_serverWebSocket);
+                if (_clientWebSocketTcs.TrySetResult(websockets.Item1))
+                {
+                    Context.HttpContext.Response.StatusCode = 101; // Switching Protocols
+                    _serverWebSocket = websockets.Item2;
+                    return Task.FromResult<WebSocket>(_serverWebSocket);
+                }
+                else
+                {
+                    Context.HttpContext.Response.StatusCode = 500; // Internal Server Error
```

Comment:
Any non-101 response is considered a failed handshake. Given that we do not know the reason that the cancellation token is canceled, I think 500 is appropriate here.